lib/livebook/apps/deployer.ex

```diff
@@ -0,0 +1,186 @@
+
+  ## Options
+
+    * `:start_only` - when `true`, deploys only if the app does not
+      exist already. If the app does exist, the deployment finishes
+      with an error. Defaults to `false`
+
+    * `:permanent` - whether the app is deployed as permanent. Defaults
+      to `false`
+
+  """
+  @spec deploy_monitor(pid(), Apps.AppSpec.t(), keyword()) :: reference()
+  def deploy_monitor(pid, app_spec, opts \\ []) do
+    opts = Keyword.validate!(opts, start_only: false, permanent: false)
+
+    ref = Process.monitor(pid)
+    GenServer.cast(pid, {:deploy, app_spec, opts[:start_only], opts[:permanent], self(), ref})
+    ref
+  end
+
+  @doc """
+  Returns a list of all deployers present in the cluster.
+  """
+  @spec list_deployers() :: list(pid())
+  def list_deployers() do
+    :pg.get_members(Apps.Deployer.PG, Apps.Deployer)
+  end
+
+  @doc """
+  Returns a node-local deployer.
+  """
+  @spec local_deployer(node()) :: pid()
+  def local_deployer(node \\ node()) do
+    list_deployers() |> Enum.find(&(node(&1) == node)) || raise "no local deployer running"
+  end
+
+  @impl true
+  def init({}) do
+    :pg.join(Apps.Deployer.PG, __MODULE__, self())
+
+    {:ok, %{}}
+  end
+
+  @impl true
+  def handle_cast({:deploy, app_spec, start_only, permanent, from, ref}, state) do
+    Logger.info("[app=#{app_spec.slug}] Deploying app")
+
+    files_tmp_path = Apps.generate_files_tmp_path(app_spec.slug)
+
+    result =
+      with {:ok, %{notebook: notebook, warnings: warnings}} <-
+             Apps.AppSpec.load(app_spec, files_tmp_path) do
+        if Apps.AppSpec.should_warmup?(app_spec) do
+          with {:error, message} <- Apps.warmup_app(notebook, files_tmp_path) do
+            Logger.warning("[app=#{app_spec.slug}] App warmup failed, #{message}")
+          end
+        end
+
+        deployment_bundle = %{
+          notebook: notebook,
+          files_tmp_path: files_tmp_path,
+          app_spec: app_spec,
+          permanent: permanent,
+          warnings: warnings
+        }
+
+        name = Apps.global_name(app_spec.slug)
+        opts = [start_only: start_only]
+
+        with {:error, error} <- start_or_redeploy(name, deployment_bundle, opts) do
+          File.rm_rf(files_tmp_path)
+          {:error, error}
+        end
+      end
+
+    send(from, {:deploy_result, ref, result})
+
+    {:noreply, state}
+  end
+
+  defp start_or_redeploy(name, deployment_bundle, opts) do
+    case :global.whereis_name(name) do
+      :undefined ->
+        with {:ok, pid} <- start_app(deployment_bundle) do
+          # We could use a global transaction to prevent the unlikely
+          # case of multiple nodes starting the app simultaneously.
+          # However, the registration can still fail if the node joins
+          # the cluster while the app process is starting. So we handle
+          # both cases the same way in the failure brach.
+          case :global.register_name(name, pid, &resolve_app_conflict/3) do
```

Comment:
Why not have the app track itself? By passing {:global, _} as a name and then checking if it returns {:already_started, pid}?